File: medium/102_Binary_Tree_Level_Order_Traversal.cpp
#include "../libstruct.h"

using namespace std;

/*
 * NO.102
 * Given a binary tree, return the level order traversal of its 
 * nodes' values. (ie, from left to right, level by level).
 */

class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        buildVector(root, 0);
        return ret;
    }
private:
    vector<vector<int> > ret;

    void buildVector(TreeNode *root, int depth)
    {
        if(root == NULL) return;
        if(ret.size() == depth)
            ret.push_back(vector<int>());

        ret[depth].push_back(root->val);
        buildVector(root->left, depth + 1);
        buildVector(root->right, depth + 1);
    }
};

/* 
 * [3,9,20,null,null,15,7]
 * return
 *  [
 *    [3],
 *    [9,20],
 *    [15,7]
 *  ]
 */
TreeNode* testTree()
{
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    return root;
}

int main(int argc, char const *argv[])
{
    Solution solution;
    vector<vector<int> > valList = solution.levelOrder(testTree());
    for (auto vec : valList)
    {
        for (auto val : vec)
        {
            cout << val << " ";
        }
        cout << endl;
    }
    return 0;
}

File: medium/102_Binary_Tree_Level_Order_Traversal.swift
//
//  102_Binary_Tree_Level_Order_Traversal.swift
//  LeetCode
//
//  Created by wangzhe on 2018/7/9.
//  Copyright © 2018年 wangzhe. All rights reserved.
//

import Foundation

/*
 * NO.102
 * Given a binary tree, return the level order traversal of its nodes' values.
 * (ie, from left to right, level by level).
 */

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

class Solution {
    func levelOrder(_ root: TreeNode?) -> [[Int]] {
        var result = [[Int]]()
        if root == nil {
            return result
        }
        var queue = [TreeNode]()
        queue.append(root!)
        while !queue.isEmpty {
            let temp = queue
            var res = [Int]()
            queue.removeAll()
            for node in temp {
                res.append(node.val)
                if let left = node.left {
                    queue.append(left)
                }
                if let right = node.right {
                    queue.append(right)
                }
            }
            result.append(res)
        }
        return result
    }
}
